### PR TITLE
Framework: Try upgrading to webpack2

### PR DESCRIPTION
**Pull request by ehg**
_WIP_

Main problems:
 - ~~We've lost the loading section dot/state~~ 
 - Using cache-identifier with babel-loader causes errors
 - ~~We'll need a Promise polyfill for IE11, if we don't already have one~~ — I think the runtime transform may take care of this
 - …

TODO:
 - Fix problems
 - Renable `react-hot-loader`
 - Test/cleanup
 - Codemod `requires` into imports, in a separate PR
 - Enable tree shaking

To test:
 - `make distclean run`

### CODE REVIEW

File: webpack.config.js [hunk 1]
```diff
@@ -12,7 +12,7 @@ const webpack = require( 'webpack' ),
  */
 const config = require( './server/config' ),
 	sections = require( './client/sections' ),
-	cacheIdentifier = require( './server/bundler/babel/babel-loader-cache-identifier' ),
+//	cacheIdentifier = require( './server/bundler/babel/babel-loader-cache-identifier' ),
```
Comment (ehg) [inline webpack.config.js]: When uncommented:
```
Error: Cannot find module '/Users/chris/a8c/wp-calypso/node_modules/babel-loader/lib/package'
    at Function.Module._resolveFilename (module.js:469:15)
    at Function.Module._load (module.js:417:25)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at getModuleVersion (/Users/chris/a8c/wp-calypso/server/bundler/babel/babel-loader-cache-identifier/index.js:14:9)
    at Object.<anonymous> (/Users/chris/a8c/wp-calypso/server/bundler/babel/babel-loader-cache-identifier/index.js:26:18)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
make: *** [run] Error 1
```

Comment (ehg) [inline webpack.config.js]: f29a4aa :(

File: server/bundler/plugin.js [hunk 2]
```diff
@@ -30,11 +31,12 @@ ChunkFileNames.prototype.apply = function( compiler ) {
 						"script.onerror = script.onload = script.onreadystatechange = null;",
 						"delete installedChunks[ chunkId ];",
 						"window.__chunkErrors[ " + JSON.stringify( chunkMaps.name ) + "[chunkId]||chunkId ]=new Error();",
-						"callback.call( null, " + this.requireFn + ")"
+						"return Promise().resolve();"
```
Comment (blowery) [inline server/bundler/plugin.js]: do you need the parens here?

Comment (ehg) [inline server/bundler/plugin.js]: Oops, nope. Fixed in de3c00f8d along with some more tidying. I wonder if we should just convert this unreadable nonsense to template literal syntax.

Comment (blowery) [inline server/bundler/plugin.js]: Indeed. Or see if we can ditch our custom loader altogether.

Comment (ehg) [inline server/bundler/plugin.js]: > Or see if we can ditch our custom loader altogether.

You mean this plugin, or https://github.com/Automattic/wp-calypso/blob/master/server/bundler/loader.js ?

We might be able to ditch this plugin, if we can look for errors in the returned rejected promise from `require.ensure` in `bundler/loader` — which handles retry logic atm using the custom global `__chunkErrors`.

Another difference seems to be https://github.com/Automattic/wp-calypso/pull/10799/files#diff-af0ff8f84f0c14d14267bd8ccd37db5fR41 — where we're serving minified assets, depending on debug environment. We could likely just minify in webpack instead.

Comment (blowery) [inline server/bundler/plugin.js]: I meant this one. Yeah, it does the retry stuff... Not sure how to replace that. 

https://github.com/Automattic/wp-calypso/pull/10634 teaches webpack to minimize our JS, which has other benefits. We'd lose the ability to serve the unminified version easily, but if we turn on sourcemaps for production-y environments, it's not that much of a loss.

**Comment by aduth** [comment]
To take advantage of tree-shaking, we'll need to update the Babel configuration to not have it apply module transformations. There's an option in the ES2015 preset to disable this, `"modules": false`, e.g.:

https://github.com/aduth/dones/blob/12c41c4/.babelrc#L5

I've found this can be tricky in SSR projects where depending on context you _do_ want modules to be transformed (Webpack can interpret `import` but Node cannot yet). This probably applies to us. In these cases I've found this project can be a good substitute for the ES2015 preset:

https://www.npmjs.com/package/babel-preset-es2015-maybe-webpack

**Comment by ehg** [comment]
> To take advantage of tree-shaking, we'll need to update the Babel configuration to not have it apply module transformations. There's an option in the ES2015 preset to disable this, "modules": false, 

Yep, I'm thinking we defer this to a separate PR, as there are other problems, like the default exports module not working with webpack2/tree shaking. Also the way we do exports in some places may be fundamentally incompatible with tree shaking? I'll post some more details on my findings tomorrow.

We also need to get rid of react-hot-loader1, as it won't work with webpack2 & `modules: false`.

> I've found this can be tricky in SSR projects where depending on context you do want modules to be transformed (Webpack can interpret import but Node cannot yet). This probably applies to us. In these cases I've found this project can be a good substitute for the ES2015 preset:
> https://www.npmjs.com/package/babel-preset-es2015-maybe-webpack

Heh, yep, I'm using this in some commits I haven't pushed. Will sort through what I've got tomorrow.

**Comment by alisterscott** [comment]
Closing as this PR is stale - please reopen if necessary

**Pull request closed without merging**